numpyro/contrib/control_flow/scan.py

```diff
@@ -0,0 +1,176 @@
+
+    if value is not None:
+        value_ndim = jnp.ndim(value)
+        sample_shape = site['kwargs']['sample_shape']
+        fn_ndim = len(sample_shape + site['fn'].shape())
+        if value_ndim == fn_ndim:
+            # this branch happens when substitute_fn is init_strategy,
+            # where we apply init_strategy to each element in the scanned series
+            return value
+        elif value_ndim == fn_ndim + 1:
+            # this branch happens when we substitute a series of values
+            shape = jnp.shape(value)
+            if shape[0] == length:
+                return value[i]
+            elif shape[0] < length:
+                rng_key = site['kwargs']['rng_key']
+                assert rng_key is not None
+                # we use the substituted values if i < shape[0]
+                # and generate a new sample otherwise
+                return lax.cond(i < shape[0],
+                                (value, i),
+                                lambda val: val[0][val[1]],
+                                rng_key,
+                                lambda val: site['fn'](rng_key=val, sample_shape=sample_shape))
+            else:
+                raise RuntimeError(f"Substituted value for site {site['name']} "
+                                   "requires length greater than or equal to scan length."
+                                   f" Expected length >= {length}, but got {shape[0]}.")
+        else:
+            raise RuntimeError(f"Something goes wrong. Expected ndim = {fn_ndim} or {fn_ndim+1},"
+                               f" but got {value_ndim}. Please report the issue to us!")
+
+
+def scan_wrapper(f, init, xs, length, reverse, rng_key=None, substitute_stack=[]):
+
+    def body_fn(wrapped_carry, x):
+        i, rng_key, carry = wrapped_carry
+        rng_key, subkey = random.split(rng_key) if rng_key is not None else (None, None)
+
+        with handlers.block():
+            seeded_fn = handlers.seed(f, subkey) if subkey is not None else f
+            for subs_type, subs_map in substitute_stack:
+                subs_fn = partial(_subs_wrapper, subs_map, i, length)
+                if subs_type == 'condition':
+                    seeded_fn = handlers.condition(seeded_fn, condition_fn=subs_fn)
+                elif subs_type == 'substitute':
+                    seeded_fn = handlers.substitute(seeded_fn, substitute_fn=subs_fn)
+
+            with handlers.trace() as trace:
+                carry, y = seeded_fn(carry, x)
+
+        return (i + 1, rng_key, carry), (PytreeTrace(trace), y)
+
+    if length is None:
+        length = tree_flatten(xs)[0][0].shape[0]
+    return lax.scan(body_fn, (jnp.array(0), rng_key, init), xs, length=length, reverse=reverse)
+
+
+def scan(f, init, xs, length=None, reverse=False):
+    """
+    This primitive scans a function over the leading array axes of
+    `xs` while carrying along state. See :func:`jax.lax.scan` for more
+    information.
```

Comment:
I think the current limitation is:
+ It only records `sample` and `deterministic` sites in `f`, not `param`. The main reason is we need to have logics to flatten/unflatten `constraint` field and I was lazy to do so.
+ I think most handlers (e.g. `scale`, `mask`) will work because we do `apply_stack` for scanned messages. The one that won't work is `plate`, where users will need to move `plate` statements inside `f`. I'll add a note to it. Do you suspect something else that won't work out of the box?
+ For distributions, I have raised errors for not supported distributions and told users to report the issue to us.

Reply:
I think it is totally fine to not support `param`, to begin with. I think most effect handlers should work out of the box, but let me take a look in detail.

We can add a disclaimer.

.. warning
This is an experimental utility function that allows users to use JAX control flow with NumPyro's effect handlers. Currently, `sample` and `deterministic` sites within the scan body are supported. If you notice that any effect handlers or distributions are unsupported, please file an issue.

Reply:
Thanks Neeraj, I'll add it.